File: 49.group-anagrams.java
/*
 * @lc app=leetcode id=49 lang=java
 *
 * [49] Group Anagrams
 */
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
;

// @lc code=start
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        List<List<String>> output = new ArrayList<>();
        List<Integer> usedIndexes = new ArrayList<>();
        List<String> charArrayList = new ArrayList<>();

        // convert strs to charArrays and build charArrayList
        for (String str : strs) {
            char[] strToCharArr = str.toCharArray();
            Arrays.sort(strToCharArr);
            String backToStr = new String(strToCharArr);
            charArrayList.add(backToStr);
        }

        // start from first sorted str, loop charArrayList
        for (int index = 0; index < charArrayList.size(); index++) {
            // check if str not already used
            if (usedIndexes.contains(index)) {
                continue;
            }
            String str = strs[index];
            List<String> currentList = new ArrayList<>();
            currentList.add(str);
            usedIndexes.add(index);

            // inner loop to compare other sorted str
            for (int next = index + 1; next < charArrayList.size(); next++) {
                String nextStr = strs[next];
                String firstWord = charArrayList.get(index);
                String nextWord = charArrayList.get(next);
                // compare with first nextStr
                if (firstWord.equals(nextWord) || nextStr == null) {
                    // if matches, add to currentList
                    currentList.add(nextStr);
                    usedIndexes.add(next);
                }
            }
            output.add(currentList);
        }
        return output;
    }
}
// @lc code=end

File: 49.group-anagrams.js
/*
 * @lc app=leetcode id=49 lang=javascript
 *
 * [49] Group Anagrams
 */

// @lc code=start
/**
 * @param {string[]} strs
 * @return {string[][]}
 */
var groupAnagrams = function (strs) {
  let countMap = new Map();
  // loop each string in array of strings
  for (let str of strs) {
    let sorted = str.split("").sort().join("");
    // check of map is empty
    if (countMap.size === 0) {
      // store first map and answer array
      countMap.set(sorted, [str]);
      continue;
    }
    // check if map contains hash already
    let i = 0;
    for (let [key, value] of countMap) {
      i++;
      // if there's no match
      if (key === sorted) {
        // add hash to map
        // push str to list
        let ansArr = countMap.get(key);
        ansArr.push(str);
        countMap.set(key, ansArr);
        break;
      }
      if (i === countMap.size) {
        countMap.set(sorted, [str]);
        break;
      }
      continue;
    }
  }
  return Array.from(countMap.values());
};
// @lc code=end

// @after-stub-for-debug-begin
module.exports = groupAnagrams;
// @after-stub-for-debug-end
